### ISSUE TITLE
floating point intrinsics may clobber errno in unrelated code

### ISSUE THREAD

**Issue by thestinger** [rust-lang/rust#12695]
These intrinsics assume `errno` is not set, but it may be set. In theory, one of these intrinsics could be reordered between a call to a C function setting `errno` and a check of the `errno` value in Rust. Rust could ignore this problem, as it will rarely (if ever) occur. However, safety mandates that we either supply our own math library known to not set errno on math errors (glibc is a black sheep in this regard) or stop using these unless something like `-fno-math-errno` is passed (as `clang` handles it on Linux).

**Comment by steveklabnik** [comment]
Traige: no change.

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by DemiMarie** [comment]
What about Julia's [OpenLibM](https://github.com/JuliaLang/openlibm)?

**Comment by Mark-Simulacrum** [comment]
Nominating for lang team: Is this I-unsound? cc @aturon

**Comment by aturon** [comment]
This is not a soundness bug per se, but rather "just" incorrect codegen. P-low.

**Comment by jonas-schievink** [comment]
Miscompilations are also soundness bugs, labeling accordingly.

**Comment by Mark-Simulacrum** [comment]
I cannot find evidence today of the claim that LLVM's floating point intrinsics require errno to not be set -- indeed, looking for errno in the LangRef seems to indicate that they all guarantee that errno would not be set (or make no mention of it).

OTOH, it does seem true that if we're lowering to libm or otherwise, errno can be set -- and that may mask a later access to errno (as mentioned, due to reordering or just lack of knowledge). But I think to some extent I'd argue that's *not* a soundness bug, as the codegen should be correct, just not do what you expect (i.e. there should not be a memory error here, though there could be a logic bug).

**Comment by hanna-kruppe** [comment]
Yeah, I don't think this is a soundness issue. It just means we effectively don't support interacting with C `errno` from Rust (at least not while using these math intrinsics anywhere in the general vicinity).

**Comment by y21** [comment]
(That PR should not have closed this issue. It updates a submodule with commits that reference another repo's issue that happens to have the same issue number as this one...)

**Comment by hkBst** [comment]
@beetrees your example now fails to reproduce on the playground...

**Comment by ds84182** [comment]
It's quite surprising that some supposedly pure intrinsics are not treated as such anymore, at least for LICM.

I think this deserves a note in the documentation for all functions that may call libm functions that _do_ modify errno (e.g. most of them except `fma`). Something like "It is unspecified whether `errno` is set when passed an [erroneous input](https://man7.org/linux/man-pages/man3/fmod.3.html#ERRORS). For example, target architecture, operating system, target feature flags, optimization level, constant propagation, and libm implementation may influence this behavior."

Rust could also guard calls to libm functions to avoid triggering math errors.

**Comment by beetrees** [comment]
> @beetrees your example now fails to reproduce on the playground...

A small tweak gets LLVM to still perform the optimisation:

```rust
#[inline(never)]
fn some_c_function() {
	// This could be any function that sets errno. For demonstration purposes, set errno directly.
	// This is the function used to get a pointer to errno on Linux, see the URL below for other platforms:
	// https://github.com/rust-lang/rust/blob/aa877bc71c8c8082122bee23d17c8669f30f275d/library/std/src/sys/pal/unix/os.rs#L52
	unsafe extern "C" {
		fn __errno_location() -> *mut i32;
	}
	unsafe { *__errno_location() = 1; }
}

fn main() {
	func(std::hint::black_box(f64::INFINITY));
}

#[inline(never)]
fn func(inf: f64) {
    some_c_function();
    // LLVM will hoist `inf % 1.0` to here.
    loop {
    	// This should display an error code of 1 ("Operation not permitted"), but will display error
    	// code 33 ("Numerical argument out of domain") instead when compiled with optimisations.
        dbg!(std::io::Error::last_os_error());
        if std::hint::black_box(true) {
            break;
        }
        std::hint::black_box(inf % 1.0);
    }
}
```

I've tested this locally (on `x86_64-unknown-linux-gnu`) and it still shows error code 33, however on the playground the linker seems to currently seems favour [the implementation from the Rust `compiler-builtins`](https://github.com/rust-lang/compiler-builtins/blob/89d20d9352892609d726ebb981fb53cadb381d1a/libm/src/math/fmod.rs), which does not set `errno` (`compiler-builtins` supplies an implementation on all non-Apple non-Windows platforms).
